File: composeApp/src/androidMain/kotlin/com/sixclassguys/maplecalendar/ui/component/AlarmSettingDialog.kt
package com.sixclassguys.maplecalendar.ui.component

import android.annotation.SuppressLint
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.focusable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.PointerEventPass
import androidx.compose.ui.input.pointer.PointerEventType
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalFocusManager
import androidx.compose.ui.platform.LocalSoftwareKeyboardController
import androidx.compose.ui.platform.LocalView
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import com.sixclassguys.maplecalendar.domain.model.MapleEvent
import com.sixclassguys.maplecalendar.theme.MapleBlack
import com.sixclassguys.maplecalendar.theme.MapleGray
import com.sixclassguys.maplecalendar.theme.MapleOrange
import com.sixclassguys.maplecalendar.theme.MapleStatBackground
import com.sixclassguys.maplecalendar.theme.MapleStatTitle
import com.sixclassguys.maplecalendar.theme.MapleWhite
import com.sixclassguys.maplecalendar.utils.minusMonths
import com.sixclassguys.maplecalendar.utils.plusMonths
import kotlinx.datetime.Clock
import kotlinx.datetime.DatePeriod
import kotlinx.datetime.LocalDate
import kotlinx.datetime.LocalDateTime
import kotlinx.datetime.TimeZone
import kotlinx.datetime.plus
import kotlinx.datetime.todayIn

@SuppressLint("MutableCollectionMutableState")
@OptIn(ExperimentalLayoutApi::class)
@Composable
fun AlarmSettingDialog(
    event: MapleEvent,
    onDismiss: () -> Unit,
    onSubmit: (List<LocalDateTime>) -> Unit
) {
    var selectedTab by remember { mutableIntStateOf(0) } // 0: 선택, 1: 주기
    // 💡 상태 관리
    var selectedDates by remember { mutableStateOf(setOf<LocalDate>()) } // 다중 선택 날짜
    val today = remember { Clock.System.todayIn(TimeZone.currentSystemDefault()) }
    var currentMonthDate by remember { mutableStateOf(LocalDate(today.year, today.month, 1)) }

    // 최종 추가된 알림 리스트 (TreeMap 활용을 권장하셨으므로 내부적으로 정렬 유지)
    var addedAlarms by remember {
        mutableStateOf(java.util.TreeMap<LocalDateTime, Boolean>(event.notificationTimes.associateWith { true }
            .toMap()))
    }

    // 시간 선택 상태
    var hour by remember { mutableStateOf("10") }
    var minute by remember { mutableStateOf("30") }

    val focusManager = LocalFocusManager.current
    val keyboardController = LocalSoftwareKeyboardController.current
    val view = LocalView.current // 💡 현재 다이얼로그의 Native View
    val dummyFocusRequester = remember { FocusRequester() }
    val minuteFocusRequester = remember { FocusRequester() }
    val clearFocusAll = {
        // Compose 레벨에서 포커스 해제
        focusManager.clearFocus(force = true)
        // 가짜 타겟으로 포커스 강제 이동 (TextField에서 포커스 뺏기)
        dummyFocusRequester.requestFocus()
        // 네이티브 뷰 레벨에서 포커스 해제 (이게 핵심)
        view.clearFocus()
        // 키보드 숨기기
        keyboardController?.hide()
    }
    val addAlarmAction = {
        if (selectedDates.isNotEmpty()) {
            val newMap = java.util.TreeMap<LocalDateTime, Boolean>(addedAlarms)
            selectedDates.forEach { date ->
                val h = hour.toIntOrNull() ?: 0
                val m = minute.toIntOrNull() ?: 0
                newMap[LocalDateTime(date.year, date.month, date.dayOfMonth, h, m)] = true
            }
            addedAlarms = newMap
            selectedDates = emptySet()
        }
        clearFocusAll()
    }

    var selectedInterval by remember { mutableIntStateOf(1) } // 기본값 '매일'
    // 💡 주기 모드용 알림 생성 로직
    val applyPeriodAlarms = {
        val h = hour.toIntOrNull() ?: 0
        val m = minute.toIntOrNull() ?: 0

        // 🚀 수정 포인트: 이벤트 시작일이 오늘보다 이전이면 '오늘'부터 계산 시작
        val calculationStart = if (event.startDate < today) today else event.startDate

        val newMap = java.util.TreeMap<LocalDateTime, Boolean>(addedAlarms)

        // 1. 계산 시작일부터 종료일까지 주기적으로 추가
        var current = calculationStart
        while (current <= event.endDate) {
            newMap[LocalDateTime(current.year, current.month, current.dayOfMonth, h, m)] = true
            current = current.plus(DatePeriod(days = selectedInterval))
        }

        // 2. 종료일 알람은 주기에 상관없이 무조건 포함 (TreeMap이 중복 자동 제거)
        val endDateTime = LocalDateTime(
            event.endDate.year, event.endDate.month, event.endDate.dayOfMonth, h, m
        )
        newMap[endDateTime] = true

        addedAlarms = newMap // 기존 리스트 초기화 후 교체
        clearFocusAll()
    }

    Dialog(
        onDismissRequest = onDismiss,
        properties = DialogProperties(usePlatformDefaultWidth = false)
    ) {
        Box(
            modifier = Modifier
                .fillMaxSize()
                .pointerInput(Unit) {
                    // 이 부분이 'Initial' 패스를 사용하여 자식보다 먼저 터치를 감지합니다.
                    awaitPointerEventScope {
                        while (true) {
                            val event = awaitPointerEvent(PointerEventPass.Initial)
                            if (event.type == PointerEventType.Press) {
                                clearFocusAll()
                            }
                        }
                    }
                },
            contentAlignment = Alignment.Center
        ) {
            Box(
                modifier = Modifier
                    .size(0.dp)
                    .focusRequester(dummyFocusRequester)
                    .focusable()
            )

            Surface(
                modifier = Modifier
                    .fillMaxWidth(0.9f)
                    .clickable(
                        interactionSource = remember { MutableInteractionSource() },
                        indication = null
                    ) { clearFocusAll() }, // 🚀 몸체 클릭 시에도 실행
                shape = RoundedCornerShape(16.dp),
                color = MapleStatBackground
            ) {
                Column(
                    modifier = Modifier.padding(16.dp),
                    horizontalAlignment = Alignment.Start
                ) {
                    Text(
                        text = "ALARM SETTING",
                        style = TextStyle(
                            color = MapleStatTitle,
                            fontWeight = FontWeight.Bold,
                            fontSize = 18.sp
                        ),
                        modifier = Modifier.fillMaxWidth()
                    )

                    Spacer(modifier = Modifier.height(16.dp))

                    Surface(
                        modifier = Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(16.dp),
                        color = MapleWhite
                    ) {
                        Column(
                            modifier = Modifier
                                .padding(16.dp)
                                .clickable(
                                    interactionSource = remember { MutableInteractionSource() },
                                    indication = null
                                ) { clearFocusAll() }, // 🚀 화이트 카드 클릭 시에도 실행
                            horizontalAlignment = Alignment.CenterHorizontally
                        ) {
                            // 💡 탭 메뉴 (선택 / 주기)
                            Row(
                                modifier = Modifier
                                    .width(144.dp) // 전체 너비 제한 (작게 설정)
                                    .height(28.dp) // 높이도 조금 더 컴팩트하게 조절
                                    .align(Alignment.Start) // 부모 Column 내에서 왼쪽으로 배치
                                    .clip(RoundedCornerShape(20.dp)) // 전체를 캡슐 모양으로 깎음
                                    .background(MapleGray) // 기본 배경색 (연한 회색)
                            ) {
                                // 1. 선택 탭
                                Box(
                                    modifier = Modifier
                                        .weight(1f)
                                        .fillMaxHeight()
                                        .background(if (selectedTab == 0) MapleOrange else Color.Transparent) // 선택 시 오렌지색
                                        .clickable { selectedTab = 0 },
                                    contentAlignment = Alignment.Center
                                ) {
                                    Text(
                                        text = "선택",
                                        style = TextStyle(
                                            color = if (selectedTab == 0) MapleWhite else MapleBlack, // 선택 시 흰색, 아닐 때 검정
                                            fontWeight = FontWeight.Bold,
                                            fontSize = 16.sp
                                        )
                                    )
                                }

                                // 2. 주기 탭
                                Box(
                                    modifier = Modifier
                                        .weight(1f)
                                        .fillMaxHeight()
                                        .background(if (selectedTab == 1) MapleOrange else Color.Transparent) // 선택 시 오렌지색
                                        .clickable { selectedTab = 1 },
                                    contentAlignment = Alignment.Center
                                ) {
                                    Text(
                                        text = "주기",
                                        style = TextStyle(
                                            color = if (selectedTab == 1) MapleWhite else MapleBlack, // 선택 시 흰색, 아닐 때 검정
                                            fontWeight = FontWeight.Bold,
                                            fontSize = 16.sp
                                        )
                                    )
                                }
                            }

                            Spacer(modifier = Modifier.height(16.dp))

                            // 💡 모드별 컨텐츠
                            if (selectedTab == 0) {
                                // 선택 모드: 실제 구현 시에는 Calendar 라이브러리나 DatePicker 배치
                                CalendarPlaceholder(
                                    startDate = if (event.startDate < today) today else event.startDate,
                                    endDate = event.endDate,
                                    currentMonthDate = currentMonthDate,
                                    selectedDates = selectedDates,
                                    onDateClick = { date ->
                                        selectedDates = if (selectedDates.contains(date)) {
                                            selectedDates - date
                                        } else {
                                            selectedDates + date
                                        }
                                    },
                                    onMonthChange = { offset ->
                                        // 월 변경 로직 (단순화된 예시)
                                        currentMonthDate = if (offset > 0) {
                                            currentMonthDate.plusMonths(1)
                                        } else {
                                            currentMonthDate.minusMonths(1)
                                        }
                                    }
                                )
                            } else {
                                // 주기 모드: 드롭다운 메뉴
                                PeriodSelector(
                                    onPeriodSelected = { selectedInterval = it }
                                )
                            }

                            Spacer(modifier = Modifier.height(16.dp))

                            TimeInputRow(
                                hour = hour, onHourChange = { hour = it },
                                minute = minute, onMinuteChange = { minute = it },
                                isAddEnabled = ((selectedTab == 0) && selectedDates.isNotEmpty()) || (selectedTab == 1), // 👈 날짜 선택 여부 체크
                                onAddClick = {
                                    if (selectedTab == 0) {
                                        addAlarmAction()
                                    } else {
                                        applyPeriodAlarms()
                                    }
                                },
                                focusRequester = minuteFocusRequester,
                                onNext = { minuteFocusRequester.requestFocus() },
                                onDone = { addAlarmAction() }
                            )

                            Spacer(modifier = Modifier.height(16.dp))

                            // 💡 알림 시간 추가 리스트 (FlowRow로 태그 형태 구현)
                            Text(
                                text = "알림 시간 추가",
                                color = MapleWhite,
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .padding(bottom = 8.dp),
                                style = TextStyle(fontWeight = FontWeight.Bold, fontSize = 14.sp)
                            )
                            Row(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .height(48.dp) // Chip 높이에 맞게 적절히 조절
                                    .horizontalScroll(rememberScrollState()), // 🚀 수평 스크롤 활성화
                                horizontalArrangement = Arrangement.spacedBy(8.dp),
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                addedAlarms.keys.forEach { alarm ->
                                    AlarmChip(
                                        alarm = alarm,
                                        onRemove = {
                                            val newMap = java.util.TreeMap<LocalDateTime, Boolean>(
                                                addedAlarms
                                            )
                                            newMap.remove(alarm)
                                            addedAlarms = newMap
                                        }
                                    )
                                }
                            }

                            Spacer(modifier = Modifier.height(24.dp))

                            // 💡 제출 버튼
                            Button(
                                onClick = { onSubmit(addedAlarms.keys.toList()) },
                                modifier = Modifier.fillMaxWidth(),
                                colors = ButtonDefaults.buttonColors(containerColor = MapleOrange),
                                shape = RoundedCornerShape(8.dp)
                            ) {
                                Text("제출하기", fontSize = 18.sp, fontWeight = FontWeight.Bold)
                            }
                        }
                    }
                }
            }
        }
    }
}

File: iosApp/iosApp/presentation/component/AlarmSettingDialog.swift
import SwiftUI
import shared

struct AlarmSettingDialog: View {
    
    let event: MapleEvent
    var onDismiss: () -> Void
    var onSubmit: ([Date]) -> Void

    @State private var selectedTab = 0 // 0: 선택, 1: 주기
    @State private var selectedDates: Set<Date> = []
    @State private var addedAlarms: [Date] = []
    
    // 달력 상태 관리
    @State private var currentMonthDate = {
        let now = Date() // 2026년 1월 20일
        let components = Calendar.current.dateComponents([.year, .month], from: now)
        return Calendar.current.date(from: components)!
    }()
    
    // 시간 입력 상태
    @State private var hour: String = "10"
    @State private var minute: String = "30"
    @State private var selectedInterval: Int = 1
    
    @FocusState private var focusedField: TimeField?
    enum TimeField { case hour, minute }

    private let today = Calendar.current.startOfDay(for: Date())

    var body: some View {
        
        ZStack {
            // 배경 딤 처리
            Color.black.opacity(0.4)
                .ignoresSafeArea()
                .onTapGesture { focusedField = nil }

            VStack(spacing: 16) {
                HStack {
                    Text("ALARM SETTING")
                        .font(.system(size: 18, weight: .bold))
                        .foregroundColor(Color.mapleStatTitle)
                    
                    Spacer()
                    
                    // 닫기 버튼 추가
                    Button(action: onDismiss) {
                        Image(systemName: "xmark")
                            .font(.system(size: 20, weight: .bold))
                            .foregroundColor(.gray)
                            .padding(4)
                    }
                }
                .frame(maxWidth: .infinity)

                VStack(spacing: 16) {
                    // 1. 탭 메뉴
                    HStack(spacing: 0) {
                        tabButton(title: "선택", index: 0)
                        tabButton(title: "주기", index: 1)
                    }
                    .background(Color.mapleGray)
                    .clipShape(Capsule())
                    .frame(width: 144, height: 32)
                    .frame(maxWidth: .infinity, alignment: .leading)

                    // 2. 컨텐츠 영역 (달력 또는 주기 선택)
                    if selectedTab == 0 {
                        CalendarViewPlaceholder(
                            startDate: max(event.startDate.toDate(), today),
                            endDate: event.endDate.toDate(),
                            currentMonthDate: $currentMonthDate,
                            selectedDates: $selectedDates,
                            onDateClick: { date in
                                if selectedDates.contains(date) {
                                    selectedDates.remove(date)
                                } else {
                                    selectedDates.insert(date)
                                }
                            }
                        )
                    } else {
                        PeriodViewPlaceholder(selectedInterval: $selectedInterval)
                    }

                    // 3. 시간 입력 및 추가 버튼
                    TimeInputRow(
                        hour: $hour,
                        minute: $minute,
                        onAddClick: addAlarmAction,
                        isAddEnabled: isAddEnabled,
                        focusedField: _focusedField
                    )

                    // 4. 알림 시간 칩 리스트
                    VStack(alignment: .leading, spacing: 8) {
                        Text("알림 시간 추가")
                            .font(.system(size: 14, weight: .bold))
                            .foregroundColor(.gray)
                        
                        ScrollView(.horizontal, showsIndicators: false) {
                            HStack(spacing: 8) {
                                ForEach(addedAlarms, id: \.self) { alarm in
                                    AlarmChip(date: alarm) { removeAlarm(alarm) }
                                }
                            }
                        }
                        .frame(height: 40)
                    }

                    // 5. 제출 버튼
                    Button(action: { onSubmit(addedAlarms) }) {
                        Text("제출하기")
                            .font(.system(size: 18, weight: .bold))
                            .foregroundColor(.white)
                            .frame(maxWidth: .infinity)
                            .padding(.vertical, 12)
                            .background(Color.mapleOrange)
                            .cornerRadius(8)
                    }
                }
                .padding(16)
                .background(Color.white)
                .cornerRadius(16)
            }
            .padding(20)
            .background(Color.mapleStatBackground)
            .cornerRadius(20)
            .padding(.horizontal, 20)
        }
    }
}

private extension AlarmSettingDialog {
    
    var isAddEnabled: Bool {
        if selectedTab == 0 { return !selectedDates.isEmpty }
        return true
    }

    func tabButton(title: String, index: Int) -> some View {
        Text(title)
            .font(.system(size: 14, weight: .bold))
            .foregroundColor(selectedTab == index ? .white : .black)
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .background(selectedTab == index ? Color.mapleOrange : Color.clear)
            .onTapGesture {
                selectedTab = index
                focusedField = nil
            }
    }

    func addAlarmAction() {
        let h = Int(hour) ?? 0
        let m = Int(minute) ?? 0
        var newAlarms = Set(addedAlarms)
        
        if selectedTab == 0 {
            for date in selectedDates {
                if let finalDate = combine(date: date, hour: h, minute: m) {
                    newAlarms.insert(finalDate)
                }
            }
            selectedDates.removeAll()
        } else {
            let start = max(event.startDate.toDate(), today)
            var current = start
            while current <= event.endDate.toDate() {
                if let finalDate = combine(date: current, hour: h, minute: m) {
                    newAlarms.insert(finalDate)
                }
                current = Calendar.current.date(byAdding: .day, value: selectedInterval, to: current) ?? current
            }
            if let endAlarm = combine(date: event.endDate.toDate(), hour: h, minute: m) {
                newAlarms.insert(endAlarm)
            }
        }
        addedAlarms = Array(newAlarms).sorted()
        focusedField = nil
    }

    func removeAlarm(_ date: Date) {
        addedAlarms.removeAll { $0 == date }
    }

    func combine(date: Date, hour: Int, minute: Int) -> Date? {
        Calendar.current.date(bySettingHour: hour, minute: minute, second: 0, of: date)
    }
}

struct AlarmChip: View {
    
    let date: Date
    let onRemove: () -> Void
    
    var body: some View {
        HStack(spacing: 4) {
            Text(formatDateTime(date))
            Button(action: onRemove) {
                Image(systemName: "xmark.circle.fill")
                    .font(.system(size: 12))
            }
        }
        .padding(.horizontal, 10)
        .padding(.vertical, 6)
        .background(Color.mapleOrange)
        .foregroundColor(.mapleWhite)
        .clipShape(Capsule())
    }
}

struct PeriodViewPlaceholder: View {
    
    @Binding var selectedInterval: Int
    private let options = [("매일", 1), ("이틀마다", 2), ("사흘마다", 3), ("일주일마다", 7)]

    var body: some View {
        
        VStack(alignment: .leading, spacing: 8) {
            Text("날짜").font(.system(size: 14, weight: .bold))
            Menu {
                ForEach(options, id: \.1) { label, value in
                    Button(label) { selectedInterval = value }
                }
            } label: {
                HStack {
                    Text(options.first { $0.1 == selectedInterval }?.0 ?? "")
                        .foregroundColor(.black)
                    Spacer()
                    Image(systemName: "chevron.down").foregroundColor(.mapleOrange)
                }
                .padding(.horizontal, 12)
                .frame(maxWidth: .infinity).frame(height: 40)
                .background(RoundedRectangle(cornerRadius: 8).stroke(Color.mapleGray))
            }
        }
    }
}
